lib/segment/tests/integration/filtrable_hnsw_test.rs

```diff
@@ -16,11 +17,64 @@ use segment::types::{
     Condition, Distance, FieldCondition, Filter, HnswConfig, Indexes, Payload, PayloadSchemaType,
     Range, SearchParams, SegmentConfig, SeqNumberType, VectorDataConfig, VectorStorageType,
 };
+use segment::vector_storage::query::discovery_query::{DiscoveryPair, DiscoveryQuery};
+use segment::vector_storage::query::reco_query::RecoQuery;
 use serde_json::json;
 use tempfile::Builder;
 
-#[test]
-fn test_filterable_hnsw() {
+const MAX_EXAMPLES: usize = 10;
+
+enum QueryVariant {
+    Nearest,
+    RecommendBestScore,
+    Discovery,
+}
+
+fn random_discovery_query<R: Rng + ?Sized>(rnd: &mut R, dim: usize) -> QueryVector {
+    let num_pairs: usize = rnd.gen_range(0..MAX_EXAMPLES / 2);
+
+    let target = random_vector(rnd, dim);
+
+    let pairs = (0..num_pairs)
+        .map(|_| {
+            let positive = random_vector(rnd, dim);
+            let negative = random_vector(rnd, dim);
+            DiscoveryPair { positive, negative }
+        })
+        .collect_vec();
+
+    DiscoveryQuery::new(target, pairs).into()
+}
+
+fn random_reco_query<R: Rng + ?Sized>(rnd: &mut R, dim: usize) -> QueryVector {
+    let num_examples: usize = rnd.gen_range(1..MAX_EXAMPLES / 2);
+
+    let positive = (0..num_examples)
+        .map(|_| random_vector(rnd, dim))
+        .collect_vec();
+    let negative = (0..num_examples)
+        .map(|_| random_vector(rnd, dim))
+        .collect_vec();
+
+    RecoQuery::new(positive, negative).into()
+}
+
+fn random_query<R: Rng + ?Sized>(variant: &QueryVariant, rnd: &mut R, dim: usize) -> QueryVector {
+    match variant {
+        QueryVariant::Nearest => random_vector(rnd, dim).into(),
+        QueryVariant::Discovery => random_discovery_query(rnd, dim),
+        QueryVariant::RecommendBestScore => random_reco_query(rnd, dim),
+    }
+}
+
+#[rstest]
+#[case::nearest(QueryVariant::Nearest, 5)]
+#[case::discovery(QueryVariant::Discovery, 25)]
```

Comment:
This level of precision is not great, but it is what's needed to pass the test. 

Something curious: I experimented with and without the 2-step search, and got indistinguishable changes in precision here.

Reply:
tbh, I am a bit concerned with this numbers, maybe we would need to investigate a bit deeper